File: day19a.lua
#!/usr/bin/env lua

local pats = {}

local function solve(target)
    local memo = { [#target + 1] = 1 }
    for i = #target, 1, -1 do
        local count = 0
        for _, pat in ipairs(pats) do
            local len = #pat
            if target:sub(i, i + len - 1) == pat then
                count = count + memo[i + len]
            end
        end
        memo[i] = count
    end
    return memo[1]
end

local file = io.open("day19.txt")
local sum = { 0, 0 }

for x in file:read():gmatch("%l+") do
    pats[#pats + 1] = x
end

for target in file:read("a"):gmatch("%l+") do
    local count = solve(target)
    if count ~= 0 then sum[1] = sum[1] + 1 end
    sum[2] = sum[2] + count
end

print(sum[1], sum[2])

File: day19a.c
#include <stdio.h>
#include <string.h>

char pats[512][8 + 1];
int n;

long long solve(const char* target)
{
    long long memo[60 + 1];
    memo[strlen(target)] = 1;

    for (int i = strlen(target) - 1; i >= 0; i--) {
        long long count = 0;
        for (int p = 0; p < n; p++) {
            int len = strlen(pats[p]);
            if (!strncmp(&target[i], pats[p], len)) {
                count += memo[i + len];
            }
        }
        memo[i] = count;
    }

    return memo[0];
}

int main()
{
    FILE* file = fopen("day19.txt", "r");
    long long sum = 0, sum2 = 0;

    while (1 == fscanf(file, "%[a-z],%*[ ]", pats[n])) n++;

    char target[60 + 1];
    while (1 == fscanf(file, "%s", target)) {
        long long count = solve(target);
        sum += count != 0;
        sum2 += count;
    }

    printf("%lld %lld\n", sum, sum2);
}
